### PR TITLE
Introducing SPMD API

### PR DESCRIPTION
**Pull request by mrshenli**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #536

### CODE REVIEW

File: spmd/api.py [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+                    else:
+                        value_remap[dtn] = gm.graph.node_copy(
+                            dtn, lambda n: value_remap[n]
+                        )
+
+
+        gm.graph.lint()
+        gm.graph.eliminate_dead_code()
+        if dist.get_rank() == 0:
+            gm.graph.print_tabular()
```
Comment (mrshenli) [inline spmd/api.py:283]: Hey @Chillee, the returned backward graph looks correct to me, but when I call backward, it does not launch allreduce in the graph. Do you know why?

```
opcode         name                  target                                          args                                                         kwargs          
-------------  --------------------  ----------------------------------------------  -----------------------------------------------------------  --------        
placeholder    t_1                   t_1                                             ()                                                           {}              
placeholder    primals_5             primals_5                                       ()                                                           {}              
placeholder    addmm                 addmm                                           ()                                                           {}              
placeholder    tangents_1            tangents_1                                      ()                                                           {}              
call_function  t                     aten.t.default                                  (t_1,)                                                       {}              
call_function  mm_3                  aten.mm.default                                 (tangents_1, t)                                              {}              
call_function  t_9                   aten.t.default                                  (tangents_1,)                                                {}              
call_function  mm_4                  aten.mm.default                                 (t_9, addmm)    
call_function  t_10                  aten.t.default                                  (mm_4,)                                                      {}     [73/1873]
call_function  sum_3                 aten.sum.dim_IntList                            (tangents_1, [0], True)                                      {}              
call_function  view_2                aten.view.default                               (sum_3, [10])                                                {}              
call_function  t_11                  aten.t.default                                  (t_10,)                                                      {}              
call_function  t_12                  aten.t.default                                  (mm_3,)                                                      {}              
call_function  mm_5                  aten.mm.default                                 (t_12, primals_5)                                            {}              
call_function  t_13                  aten.t.default                                  (mm_5,)                                                      {}              
call_function  sum_4                 aten.sum.dim_IntList                            (mm_3, [0], True)                                            {}              
call_function  view_3                aten.view.default                               (sum_4, [10])                                                {}              
call_function  t_14                  aten.t.default                                  (t_13,)                                                      {}              
call_function  clone                 aten.clone.default                              (t_14,)                                                      {}              
get_attr       _tensor_constant0     _tensor_constant0                               ()                                                           {}              
get_attr       _tensor_constant1     _tensor_constant1                               ()                                                           {}              
call_function  allreduce__default    c10d.allreduce_.default                         ([clone], _tensor_constant0, _tensor_constant1, -1)          {}              
call_function  comm_result           <function _wrap_comm_result at 0x7f1e55865ee0>  (allreduce__default,)                                        {}              
call_function  getitem               <built-in function getitem>                     (comm_result, 0)                                             {}              
call_function  getitem_1             <built-in function getitem>                     (getitem, 0)                                                 {}              
call_function  wait_comm             <function _wait_comm at 0x7f1e55865b80>         (getitem_1,)                                                 {}              
call_function  clone_1               aten.clone.default                              (view_3,)                                                    {}              
get_attr       _tensor_constant0_1   _tensor_constant0                               ()                                                           {}              
get_attr       _tensor_constant1_1   _tensor_constant1                               ()                                                           {}              
call_function  allreduce__default_1  c10d.allreduce_.default                         ([clone_1], _tensor_constant0_1, _tensor_constant1_1, -1)    {}              
call_function  comm_result_1         <function _wrap_comm_result at 0x7f1e55865ee0>  (allreduce__default_1,)                                      {}              
call_function  getitem_2             <built-in function getitem>                     (comm_result_1, 0)                                           {}              
call_function  getitem_3             <built-in function getitem>                     (getitem_2, 0)                                               {}              
call_function  wait_comm_1           <function _wait_comm at 0x7f1e55865b80>         (getitem_3,)                                                 {}              
call_function  clone_2               aten.clone.default                              (t_11,)                                                      {}              
get_attr       _tensor_constant0_2   _tensor_constant0                               ()                                                           {}              
get_attr       _tensor_constant1_2   _tensor_constant1                               ()                                                           {}              
call_function  allreduce__default_2  c10d.allreduce_.default                         ([clone_2], _tensor_constant0_2, _tensor_constant1_2, -1)    {}              
call_function  comm_result_2         <function _wrap_comm_result at 0x7f1e55865ee0>  (allreduce__default_2,)                                      {}              
call_function  getitem_4             <built-in function getitem>                     (comm_result_2, 0)                                           {}              
call_function  getitem_5             <built-in function getitem>                     (getitem_4, 0)                                               {}              
call_function  wait_comm_2           <function _wait_comm at 0x7f1e55865b80>         (getitem_5,)                                                 {}
call_function  clone_3               aten.clone.default                              (view_2,)                                                    {}
get_attr       _tensor_constant0_3   _tensor_constant0                               ()                                                           {}
get_attr       _tensor_constant1_3   _tensor_constant1                               ()                                                           {}
call_function  allreduce__default_3  c10d.allreduce_.default                         ([clone_3], _tensor_constant0_3, _tensor_constant1_3, -1)    {}
call_function  comm_result_3         <function _wrap_comm_result at 0x7f1e55865ee0>  (allreduce__default_3,)                                      {}
call_function  getitem_6             <built-in function getitem>                     (comm_result_3, 0)                                           {}
call_function  getitem_7             <built-in function getitem>                     (getitem_6, 0)                                               {}
call_function  wait_comm_3           <function _wait_comm at 0x7f1e55865b80>         (getitem_7,)                                                 {}
output         output_1              output                                          ([wait_comm, wait_comm_1, wait_comm_2, wait_comm_3, None],)  {}
```

Comment (mrshenli) [reply spmd/api.py:283]: I am using functorch in pytorch/pytorch nightly `1.13.0.dev20220915+cu113`

**Review by wanchaol — COMMENTED** [review]
Great work! I know that this might still WIP, would like to discuss things earlier so added some comments.

File: spmd/api.py [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+    )
+
+    def unwrap_local(e):
+        return e._local_tensor if isinstance(e, DTensor) else e
+
+    return make_fx(dispatch)(tree_map(unwrap_local, args))
+
+
+
+class SPMD(nn.Module):
```
[the remark this answers is not in the document]
Comment (mrshenli) [reply spmd/api.py]: @aazzolini also asked for this. I extract the main logic to a `_convert_to_distributed` function API. 

Do we still plan to have a module wrapper in tau? I thought that's the original goal?

File: spmd/api.py [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+                    node, node_to_obj
+                )
+            elif node.op == "output":
+                new_args = []
+                for a in node.args[0]:
+                    if not isinstance(a, fx.Node):
+                        new_args.append(a)
+                        continue
+                    obj = node_to_obj[a]
+                    if isinstance(obj, DTensor) and isinstance(obj.placements[0], _Partial):
```
[the remark this answers is not in the document]
Comment (mrshenli) [reply spmd/api.py]: Yep, tracing train step will solve this problem. This is mainly making sure that after bw pass on an nn.Module, we are not returning partial tensors as grads.

File: spmd/api.py [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+
+import os
+from functools import partial
+from typing import Any, Callable, Dict, List, Sequence, Tuple
+
+
+@dataclass
+class Schema:
+    mesh: DeviceMesh
+    placements: List[Placement]
```
[the remark this answers is not in the document]
Comment (mrshenli) [reply spmd/api.py:38]: Yep, haven't added the `schema_override` part yet. This PR is mainly to unblock Alisson's experiment on DDP behavior + test models.

File: spmd/api.py [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+                    if not isinstance(a, fx.Node):
+                        new_args.append(a)
+                        continue
+                    obj = node_to_obj[a]
+                    if isinstance(obj, DTensor) and isinstance(
+                        obj.placements[0], _Partial
+                    ):
+                        has_partial = True
+
+                        def dummy_add(grad, zero) -> torch.Tensor:
```
Comment (mrshenli) [inline spmd/api.py:166]: Q: this code is mainly to trigger allreduce for partial -> replicate conversion, so that we don't return partial tensors after the backward pass on `nn.Module`. Let me know if we no longer want a module wrapper in tau repo

File: spmd/api.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (wanchaol) [reply spmd/api.py]: I don't have a strong opinion here, we can iterate on the APIs and finalized it later. My point was that if we are thinking a functional like API for other module wrappers like DDP, it might worth considering we just directly implement a functional like API in tau?

File: spmd/api.py (hunk 5 above)
Comment (wanchaol) [reply spmd/api.py:166]: iirc we couldn't trace a train step because aot does not consider optimizer step into its min cut algorithm? My view on this is that it might make sense to just group backward graph + optimizer graph together as lots of optimizer is done with these two together. cc @Chillee 

Let's explicitly trigger all_reduce for now to unblock, and we can discuss more what might work best.

File: test/spmd/test_tracing.py [hunk 6]
```diff
@@ -410,5 +414,65 @@ def f(x, y):
...
+        x = torch.randn(2, 10).to(device)
+        self._test_trace_replicate(model, x)
+
+    @with_comms
+    def test_parallel(self):
+        if self.device_type == "cuda":
+            device = torch.device(self.rank)
+            torch.cuda.set_device(self.rank)
+        else:
+            device = torch.device("cpu")
```
Comment (wanchaol) [inline test/spmd/test_tracing.py:460]: nit: this should all be done within the `@with_comms` wrapper, so simply `model = Model().to(self.device_type)` should work.

File: spmd/api.py [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+    # TODO: add schema_override
+    def __init__(self, module: nn.Module, schema: Schema):
+        super().__init__()
+        assert schema.placements == [
+            Replicate()
+        ], "SPMD only support Replicate() parameters for now"
+
+        # TODO: coalesce broadcasts
+        for p in module.parameters():
+            dist.broadcast(p, src=0)
```
Comment (wanchaol) [inline spmd/api.py:288]: when we replace module parameters primals in the graph, if we use `distribute_tensor(p, mesh, [Replicate()]`, it should work similarly?

**Review by aazzolini — COMMENTED** [review]
Thanks @Shen!
I patched this locally and seems like my parameters are duplicated after I run the module once. E.g.. DOing:

    f = nn.Sequential(nn.Linear(6, 10), nn.Linear(10, 2))
    x = torch.randn(4, 6)
    print(f)
    g = SPMD(f, schema=Schema(dm, [Replicate()]))
    print(g)
    g(x)
    print(g)


I get :



Sequential(
  (0): Linear(in_features=6, out_features=10, bias=True)
  (1): Linear(in_features=10, out_features=2, bias=True)
)

SPMD(
  (_local_module): Sequential(
    (0): Linear(in_features=6, out_features=10, bias=True)
    (1): Linear(in_features=10, out_features=2, bias=True)
  )
)

SPMD(
  (_local_module): Sequential(
    (0): Linear(in_features=6, out_features=10, bias=True)
    (1): Linear(in_features=10, out_features=2, bias=True)
  )
  (_compiled_m): AOTModule(
    (orig_module): Sequential(
      (0): Linear(in_features=6, out_features=10, bias=True)
      (1): Linear(in_features=10, out_features=2, bias=True)
    )
  )
)


So after compilation i get a clone. I think if i do module.parameters() this will fail?

**Comment by aazzolini** [comment]
Maybe a question for @Chillee as well, how do I get the graph from inside an AOTModule so I can print out the code?

File: spmd/api.py [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+
+        self._local_module = module
+        self._schema = schema
+        self._compiled_m = None
+
+    def _compile(
+        self, gm: fx.GraphModule, inps: List[torch.Tensor]
```
Comment (wanchaol) [inline spmd/api.py:295]: Curious are List of Tensor a strict requirement for AOT? what if user pass a single tensor in?

File: spmd/api.py [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+    ) -> fx.GraphModule:
+        def is_param(t: torch.Tensor) -> bool:
+            # N.B.: id(t) and id(param) does not match
+            return t.storage().data_ptr() in [
+                p.storage().data_ptr() for p in self._local_module.parameters()
```
Comment (wanchaol) [inline spmd/api.py:300]: Maybe we can check this by sth like `isinstance(t, nn.Parameter)`?

File: test/spmd/test_tracing.py [hunk 10]
```diff
@@ -410,5 +414,65 @@ def f(x, y):
...
+                ),
+                placements=[Replicate()],
+            ),
+        )
+
+        ddp(x).sum().backward()
+        spmd(x).sum().backward()
+        for p1, p2 in zip(ddp.parameters(), spmd.parameters()):
+            # DDP divides gradients by world size to compute average, but
+            # _Partial tensor shouldn't do that automatically. Hence explicitly
```
Comment (wanchaol) [inline test/spmd/test_tracing.py:439]: _partial don't need to do this divide because it's already working with batched sharded input?

**Review by wanchaol — APPROVED** [review]
lgtm, leave a few questions and nits.

File: spmd/api.py (hunk 9 above)
Comment (aazzolini) [reply spmd/api.py:300]: Would it be best to actually pass a list of tensors and their respective expected sharding as part of the constructor?

File: test/spmd/test_tracing.py [hunk 11]
```diff
@@ -410,5 +414,65 @@ def f(x, y):
         self._test_expand(xd, yd, f, mesh, out_placement)
 
 
+class TraceModuleTest(DistTensorTestBase):
+    @property
+    def world_size(self):
+        return 2
+
+    def _test_trace_replicate(self, model, x):
+        # if x.device.type == "cuda":
```
Comment (anj-s) [inline test/spmd/test_tracing.py:423]: nit: remove?

File: spmd/api.py [hunk 12]
```diff
@@ -0,0 +1,318 @@
...
+            ]
+
+        shard_schema = Schema(mesh=self._schema.mesh, placements=[Shard(0)])
+        schemas = [
+            self._schema if is_param(inp) else shard_schema for inp in inps
```
Comment (anj-s) [inline spmd/api.py:305]: This is ensuring that we shard input on the batch dimension for now right?

File: spmd/api.py [hunk 13]
```diff
@@ -0,0 +1,318 @@
...
+
+    gm.graph.lint()
+    gm.graph.eliminate_dead_code()
+    gm.recompile()
+
+    return gm
+
+
+class SPMD(nn.Module):
+    # TODO: add schema_override
```
Comment (anj-s) [inline spmd/api.py:279]: What is this override for?

Comment (mrshenli) [reply spmd/api.py:279]: in our original design, users can provide a function to override schema on individual parameters.

File: spmd/api.py (hunk 8 above)
Comment (mrshenli) [reply spmd/api.py:295]: I think these tensors represent the placeholders objects (params and inputs) in the gm. So, it's always a list? If there is a placeholder, I assume this will be a list of a single element?

File: spmd/api.py (hunk 9 above)
Comment (mrshenli) [reply spmd/api.py:300]: unfortunately, `t` is not an `nn.Parameter`, even `id(t)` does not match `id(p)`

File: spmd/api.py (hunk 12 above)
Comment (mrshenli) [reply spmd/api.py:305]: yep

File: test/spmd/test_tracing.py [hunk 14]
```diff
@@ -410,5 +414,65 @@ def f(x, y):
...
+        ddp = DDP(deepcopy(model))
+        spmd = SPMD(
+            deepcopy(model),
+            schema=Schema(
```
[the remark this answers is not in the document]
Comment (mrshenli) [reply test/spmd/test_tracing.py:427]: to apply the placement + device_mesh to all params.

File: test/spmd/test_tracing.py (hunk 10 above)
Comment (mrshenli) [reply test/spmd/test_tracing.py:439]: Could you please elaborate? I thought we need this because partial -> replicate triggers allreduce sum, while DDP does allreduce sum and then divide by world size. 

Whether we should do the division or not depends on the loss function. DDP chooses to apply the division because most loss function is calculating averaged per-sample loss. So dividing by world size will produce similar result as local training for the majority of use cases. 

This test is using `sum()` as loss function, so in theory DDP shouldn't divide it by world size, but however that division is hard-coded in DDP source code.

File: spmd/api.py (hunk 7 above)
Comment (mrshenli) [reply spmd/api.py:288]: Yep, but we are not replacing local param with DT yet. In this PR, DT only exist during tracing. After that, comm/comp all operates on local tensor.

**Comment by mrshenli** [comment]
Hey @aazzolini, the tests calls `.parameters()` on `AOTModule` and DDP respectively and check the values, which seems working?
